components/eventing-controller/pkg/backend/jetstreamv2/utils.go

```diff
@@ -0,0 +1,222 @@
+// Supported deliver policy values are ("all", "last", "last_per_subject" and "new").
+func toJetStreamConsumerDeliverPolicyOptOrDefault(deliverPolicy string) nats.SubOpt {
+	switch deliverPolicy {
+	case ConsumerDeliverPolicyAll:
+		return nats.DeliverAll()
+	case ConsumerDeliverPolicyLast:
+		return nats.DeliverLast()
+	case ConsumerDeliverPolicyLastPerSubject:
+		return nats.DeliverLastPerSubject()
+	case ConsumerDeliverPolicyNew:
+		return nats.DeliverNew()
+	}
+	return nats.DeliverNew()
+}
+
+func getStreamConfig(natsConfig env.NatsConfig) (*nats.StreamConfig, error) {
+	storage, err := toJetStreamStorageType(natsConfig.JSStreamStorageType)
+	if err != nil {
+		return nil, err
+	}
+	retentionPolicy, err := toJetStreamRetentionPolicy(natsConfig.JSStreamRetentionPolicy)
+	if err != nil {
+		return nil, err
+	}
+	streamConfig := &nats.StreamConfig{
+		Name:      natsConfig.JSStreamName,
+		Storage:   storage,
+		Replicas:  natsConfig.JSStreamReplicas,
+		Retention: retentionPolicy,
+		MaxMsgs:   natsConfig.JSStreamMaxMessages,
+		MaxBytes:  natsConfig.JSStreamMaxBytes,
+		// Since one stream is used to store events of all types, the stream has to match all event types, and therefore
+		// we use the wildcard char >. However, to avoid matching internal JetStream and non-Kyma-related subjects, we
+		// use a prefix. This prefix is handled only on the JetStream level (i.e. JetStream handler
+		// and EPP) and should not be exposed in the Kyma subscription. Any Kyma event type gets appended with the
+		// configured stream's subject prefix.
+		Subjects: []string{fmt.Sprintf("%s.>", env.JetStreamSubjectPrefix)},
+	}
+	return streamConfig, nil
+}
+
+func createKeyPrefix(sub *eventingv1alpha2.Subscription) string {
+	namespacedName := types.NamespacedName{
+		Namespace: sub.Namespace,
+		Name:      sub.Name,
+	}
+	return namespacedName.String()
+}
+
+func getCleanEventTypesFromStatus(subscriptionStatus eventingv1alpha2.SubscriptionStatus) []string {
+	var cleantypes []string
+	for _, eventtypes := range subscriptionStatus.Types {
+		cleantypes = append(cleantypes, eventtypes.CleanType)
+	}
+	return cleantypes
+}
+
+// TODO: to be moved to subscription types
+func getUniqueEventTypes(eventTypes []string) []string {
+	unique := make([]string, 0, len(eventTypes))
+	mapper := make(map[string]bool)
+
+	for _, val := range eventTypes {
+		if _, ok := mapper[val]; !ok {
+			mapper[val] = true
+			unique = append(unique, val)
+		}
+	}
+
+	return unique
+}
+
+// getCleanEventTypes returns a list of clean eventTypes from the unique types in the subscription.
+func getCleanEventTypes(sub *eventingv1alpha2.Subscription, cleaner cleaner.Cleaner) ([]eventingv1alpha2.EventType, error) {
+	// TODO: Put this in the validation webhook
+	if sub.Spec.Types == nil || sub.Spec.Source == "" {
+		return []eventingv1alpha2.EventType{}, errors.New("event source and types must be provided")
+	}
```

Comment:
(nit) source is not used in this method, then let's only check for types?
```
if sub.Spec.Types == nil {
	return []eventingv1alpha2.EventType{}, errors.New("event types must be provided")
}
```

Reply:
agreed! good catch